File: src/shared/display.ts
import { STATUSES, type SemanticColor } from './status'
import type { AppSnapshot, CalendarEvent, TimerState } from './types'

export interface LedContent {
  label: string
  /** Big countdown / clock line. Empty string means "no big line". */
  big: string
  sub: string
  color: SemanticColor
  /** A key that changes whenever the panel should play a mode transition. */
  transitionKey: string
}

export function pad2(value: number): string {
  return value < 10 ? `0${value}` : String(value)
}

/** `MM:SS`, or `H:MM:SS` past an hour. */
export function formatDuration(ms: number): string {
  const total = Math.max(0, Math.round(ms / 1000))
  const hours = Math.floor(total / 3600)
  const minutes = Math.floor((total % 3600) / 60)
  const seconds = total % 60
  if (hours > 0) return `${hours}:${pad2(minutes)}:${pad2(seconds)}`
  return `${pad2(minutes)}:${pad2(seconds)}`
}

/** 12-hour clock without a meridiem — `10:42`, `2:30`. */
export function formatClock(epochMs: number): string {
  const date = new Date(epochMs)
  const hours = date.getHours() % 12 || 12
  return `${hours}:${pad2(date.getMinutes())}`
}

export function formatMinutes(ms: number): string {
  return `${Math.max(0, Math.round(ms / 60000))} MIN`
}

/** "10:42 · 24 min away" nearby, an absolute weekday/time once it is hours out. */
export function describeWhen(startsAt: number, now: number): string {
  const minutesAway = Math.round((startsAt - now) / 60_000)
  if (minutesAway <= 90) {
    return `${formatClock(startsAt)} · ${minutesAway} min away`
  }
  return new Date(startsAt).toLocaleString(undefined, {
    weekday: 'short',
    hour: 'numeric',
    minute: '2-digit'
  })
}

/** Milliseconds left on the timer, computed from the deadline while running. */
export function remainingMs(timer: TimerState, now: number): number {
  if (timer.phase === 'running' && timer.endsAt !== null) {
    return Math.max(0, timer.endsAt - now)
  }
  return Math.max(0, timer.remainingMs)
}

function sanitize(text: string): string {
  return text
    .toUpperCase()
    .replace(/[^A-Z0-9 :.,\-+/'!?%()=<>*"&#~_•]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
}

/** Trim to `max` characters so long titles never overflow the panel. */
export function fitLabel(text: string, max: number): string {
  const clean = sanitize(text)
  if (clean.length <= max) return clean
  return `${clean.slice(0, Math.max(1, max - 1)).trim()}.`
}

export interface FocusSuggestion {
  freeMs: number
  suggestMinutes: number
}

/** A focus block that fits before the next event, leaving a 5 minute buffer. */
export function suggestFocus(
  next: CalendarEvent | null,
  now: number,
  presets: number[]
): FocusSuggestion | null {
  if (!next) return null
  const freeMs = next.startsAt - now
  // Only suggest when the gap is short enough that fitting a session is a real
  // decision — with hours to spare the nudge is noise.
  if (freeMs > 2 * 60 * 60_000) return null
  const usableMinutes = Math.floor(freeMs / 60000) - 5
  if (usableMinutes < 10) return null
  const candidates = [...presets].sort((a, b) => a - b).filter((p) => p <= usableMinutes)
  const suggestMinutes = candidates.length > 0 ? candidates[candidates.length - 1] : 0
  if (suggestMinutes === 0) return null
  return { freeMs, suggestMinutes }
}

function statusSub(snapshot: AppSnapshot, now: number): string {
  const { status, calendar, settings } = snapshot
  const meta = STATUSES[status.id]

  if (status.id === 'meeting') {
    const until = calendar.current?.endsAt ?? status.until
    return until ? `UNTIL ${formatClock(until)}` : 'HEADS DOWN'
  }
  if (status.id === 'dnd') {
    return status.until ? `BACK AT ${formatClock(status.until)}` : 'STAY IN THE ZONE'
  }
  if (status.id === 'oncall') return meta.sub

  // Available / custom: prefer calendar awareness, then a focus suggestion.
  if (calendar.current) {
    return fitLabel(`NOW: ${calendar.current.title} TIL ${formatClock(calendar.current.endsAt)}`, 22)
  }
  const suggestion = suggestFocus(calendar.next, now, settings.presets)
  if (calendar.next) {
    const nextLine = `NEXT: ${calendar.next.title} ${formatClock(calendar.next.startsAt)}`
    if (suggestion && suggestion.freeMs < 90 * 60000) {
      return fitLabel(
        `${formatMinutes(suggestion.freeMs)} FREE - START ${suggestion.suggestMinutes} MIN`,
        22
      )
    }
    return fitLabel(nextLine, 22)
  }
  return status.id === 'custom' ? 'READY TO FOCUS' : meta.sub
}

/** Single source of truth for what the LED panel shows. */
export function deriveLedContent(snapshot: AppSnapshot, now: number): LedContent {
  const { timer, status } = snapshot

  if (timer.phase === 'running' || timer.phase === 'paused') {
    const paused = timer.phase === 'paused'
    const isBreak = timer.mode === 'break'
    const color: SemanticColor = paused ? 'paused' : isBreak ? 'break' : 'focus'
    return {
      label: paused ? 'PAUSED' : isBreak ? 'BREAK' : 'FOCUS',
      big: formatDuration(remainingMs(timer, now)),
      sub: isBreak ? 'BREATHE & RESET' : fitLabel(timer.title || 'DEEP WORK', 22),
      color,
      transitionKey: `${timer.mode}:${timer.phase}`
    }
  }

  if (timer.phase === 'completed') {
    const isBreak = timer.mode === 'break'
    return {
      label: isBreak ? 'BREAK OVER' : 'SESSION DONE',
      big: '00:00',
      sub: isBreak ? 'BACK TO WORK' : 'TAKE A BREAK',
      color: isBreak ? 'focus' : 'break',
      transitionKey: `completed:${timer.mode}`
    }
  }

  const meta = STATUSES[status.id]
  const label =
    status.id === 'custom' ? fitLabel(status.customLabel || 'FOCUS MODE', 18) : meta.label
  return {
    label,
    big: meta.showsClock ? formatClock(now) : '',
    sub: statusSub(snapshot, now),
    color: meta.color,
    transitionKey: `status:${status.id}`
  }
}

/** LED labels are shouty by design; the menu bar is not. `ON CALL` → `On Call`. */
function titleCase(text: string): string {
  return text
    .toLowerCase()
    .replace(/(^|[\s\-/])([a-z0-9])/g, (_match, lead: string, char: string) => lead + char.toUpperCase())
}

/** Menu bar titles compete with every other app for the notch. */
const MENU_BAR_MAX = 22

/**
 * What the tray prints beside its icon. Derived from `deriveLedContent()` so the
 * menu bar can never claim a different state than the panel.
 */
export function deriveMenuBarTitle(snapshot: AppSnapshot, now: number): string {
  const mode = snapshot.settings.menuBarText
  if (mode === 'off') return ''

  const { timer } = snapshot
  const running = timer.phase === 'running'
  const active = running || timer.phase === 'paused'
  if (mode === 'timer') return running ? formatDuration(remainingMs(timer, now)) : ''

  const content = deriveLedContent(snapshot, now)
  const custom = !active && snapshot.status.id === 'custom' && snapshot.status.customLabel.trim()
  // A custom status is the user's own words — keep their capitalisation.
  const label = custom ? custom : titleCase(content.label)
  // Idle statuses show the wall clock on the panel — the menu bar already has one.
  const title = active ? `${label} ${formatDuration(remainingMs(timer, now))}` : label
  return title.length > MENU_BAR_MAX ? `${title.slice(0, MENU_BAR_MAX - 1).trim()}…` : title
}

File: apple/GerdooKit/Sources/GerdooKit/Model/Display.swift
/// The single source of truth for what the LED panel shows, ported from
/// `src/shared/display.ts`. The panel, the complications, the Live Activity and
/// the watch all read this, so they cannot disagree about the device's state.
import Foundation

public struct LedContent: Equatable, Sendable {
  public var label: String
  /// Big countdown / clock line. Empty means "no big line".
  public var big: String
  public var sub: String
  public var color: SemanticColor
  /// A key that changes whenever the panel should play a mode transition.
  public var transitionKey: String

  public init(
    label: String, big: String, sub: String, color: SemanticColor, transitionKey: String
  ) {
    self.label = label
    self.big = big
    self.sub = sub
    self.color = color
    self.transitionKey = transitionKey
  }
}

public func pad2(_ value: Int) -> String {
  value < 10 ? "0\(value)" : String(value)
}

/// `MM:SS`, or `H:MM:SS` past an hour.
public func formatDuration(_ interval: TimeInterval) -> String {
  let total = max(0, Int(interval.rounded()))
  let hours = total / 3600
  let minutes = (total % 3600) / 60
  let seconds = total % 60
  if hours > 0 { return "\(hours):\(pad2(minutes)):\(pad2(seconds))" }
  return "\(pad2(minutes)):\(pad2(seconds))"
}

/// 12-hour clock without a meridiem — `10:42`, `2:30`.
public func formatClock(_ date: Date, calendar: Calendar = .current) -> String {
  let parts = calendar.dateComponents([.hour, .minute], from: date)
  let hour = (parts.hour ?? 0) % 12
  return "\(hour == 0 ? 12 : hour):\(pad2(parts.minute ?? 0))"
}

public func formatMinutes(_ interval: TimeInterval) -> String {
  "\(max(0, Int((interval / 60).rounded()))) MIN"
}

/// "10:42 · 24 min away" nearby, an absolute weekday/time once it is hours out.
public func describeWhen(_ startsAt: Date, now: Date) -> String {
  let minutesAway = Int((startsAt.timeIntervalSince(now) / 60).rounded())
  if minutesAway <= 90 {
    return "\(formatClock(startsAt)) · \(minutesAway) min away"
  }
  return startsAt.formatted(.dateTime.weekday(.abbreviated).hour().minute())
}

/// Time left on the timer, computed from the deadline while running.
public func remaining(_ timer: TimerState, now: Date) -> TimeInterval {
  if timer.phase == .running, let endsAt = timer.endsAt {
    return max(0, endsAt.timeIntervalSince(now))
  }
  return max(0, timer.remaining)
}

/// What the panel can actually draw — anything else becomes a space.
private let ledAlphabet = Set("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 :.,-+/'!?%()=<>*\"&#~_•")

private func sanitize(_ text: String) -> String {
  let upper = text.uppercased().map { ledAlphabet.contains($0) ? $0 : " " }
  return String(upper)
    .split(separator: " ", omittingEmptySubsequences: true)
    .joined(separator: " ")
}

/// Trim to `max` characters so long titles never overflow the panel.
public func fitLabel(_ text: String, max limit: Int) -> String {
  let clean = sanitize(text)
  if clean.count <= limit { return clean }
  let cut = clean.prefix(Swift.max(1, limit - 1)).trimmingCharacters(in: .whitespaces)
  return "\(cut)."
}

public struct FocusSuggestion: Equatable, Sendable {
  public var free: TimeInterval
  public var suggestMinutes: Int
}

/// A focus block that fits before the next event, leaving a 5 minute buffer.
public func suggestFocus(
  next: CalendarEvent?, now: Date, presets: [Int]
) -> FocusSuggestion? {
  guard let next else { return nil }
  let free = next.startsAt.timeIntervalSince(now)
  // Only suggest when the gap is short enough that fitting a session is a real
  // decision — with hours to spare the nudge is noise.
  if free > 2 * 3600 { return nil }
  let usableMinutes = Int(free / 60) - 5
  if usableMinutes < 10 { return nil }
  guard let suggestMinutes = presets.sorted().last(where: { $0 <= usableMinutes }) else {
    return nil
  }
  return FocusSuggestion(free: free, suggestMinutes: suggestMinutes)
}

private func statusSub(_ snapshot: AppSnapshot, now: Date) -> String {
  let status = snapshot.status
  let calendar = snapshot.calendar
  let meta = status.id.meta

  switch status.id {
  case .meeting:
    let until = calendar.current?.endsAt ?? status.until
    return until.map { "UNTIL \(formatClock($0))" } ?? "HEADS DOWN"
  case .dnd:
    return status.until.map { "BACK AT \(formatClock($0))" } ?? "STAY IN THE ZONE"
  case .oncall:
    // A meeting that put the status here leaves its end time behind, which says
    // more than the static sub-label does.
    let until = calendar.current?.endsAt ?? status.until
    return until.map { "UNTIL \(formatClock($0))" } ?? meta.sub
  case .available, .custom:
    break
  }

  // Available / custom: prefer calendar awareness, then a focus suggestion.
  if let current = calendar.current {
    return fitLabel("NOW: \(current.title) TIL \(formatClock(current.endsAt))", max: 22)
  }
  if let next = calendar.next {
    let suggestion = suggestFocus(next: next, now: now, presets: snapshot.settings.presets)
    if let suggestion, suggestion.free < 90 * 60 {
      return fitLabel(
        "\(formatMinutes(suggestion.free)) FREE - START \(suggestion.suggestMinutes) MIN",
        max: 22)
    }
    return fitLabel("NEXT: \(next.title) \(formatClock(next.startsAt))", max: 22)
  }
  return status.id == .custom ? "READY TO FOCUS" : meta.sub
}

/// Single source of truth for what the LED panel shows.
public func deriveLedContent(_ snapshot: AppSnapshot, now: Date) -> LedContent {
  let timer = snapshot.timer
  let status = snapshot.status

  if timer.phase == .running || timer.phase == .paused {
    let paused = timer.phase == .paused
    let isBreak = timer.mode == .break
    let color: SemanticColor = paused ? .paused : (isBreak ? .break : .focus)
    return LedContent(
      label: paused ? "PAUSED" : (isBreak ? "BREAK" : "FOCUS"),
      big: formatDuration(remaining(timer, now: now)),
      sub: isBreak ? "BREATHE & RESET"
        : fitLabel(timer.title.isEmpty ? "DEEP WORK" : timer.title, max: 22),
      color: color,
      transitionKey: "\(timer.mode.rawValue):\(timer.phase.rawValue)")
  }

  if timer.phase == .completed {
    let isBreak = timer.mode == .break
    return LedContent(
      label: isBreak ? "BREAK OVER" : "SESSION DONE",
      big: "00:00",
      sub: isBreak ? "BACK TO WORK" : "TAKE A BREAK",
      color: isBreak ? .focus : .break,
      transitionKey: "completed:\(timer.mode.rawValue)")
  }

  let meta = status.id.meta
  let label: String =
    status.id == .custom
    ? fitLabel(status.customLabel.isEmpty ? "FOCUS MODE" : status.customLabel, max: 18)
    : meta.label
  return LedContent(
    label: label,
    big: meta.showsClock ? formatClock(now) : "",
    sub: statusSub(snapshot, now: now),
    color: meta.color,
    transitionKey: "status:\(status.id.rawValue)")
}

/// LED labels are shouty by design; a widget or a notification is not.
/// `ON CALL` → `On Call`.
public func titleCase(_ text: String) -> String {
  var result = ""
  var atWordStart = true
  for char in text.lowercased() {
    if atWordStart, char.isLetter || char.isNumber {
      result.append(Character(char.uppercased()))
      atWordStart = false
    } else {
      result.append(char)
      if char == " " || char == "-" || char == "/" { atWordStart = true }
    }
  }
  return result
}

/// One line for a place with no room for three — an inline complication, a
/// notification title, the Dynamic Island's compact side. Derived from
/// `deriveLedContent` so it cannot claim a state the panel disagrees with.
public func deriveShortTitle(_ snapshot: AppSnapshot, now: Date) -> String {
  let timer = snapshot.timer
  let content = deriveLedContent(snapshot, now: now)
  if timer.isActive {
    return "\(titleCase(content.label)) \(formatDuration(remaining(timer, now: now)))"
  }
  let custom = snapshot.status.customLabel.trimmingCharacters(in: .whitespaces)
  if snapshot.status.id == .custom, !custom.isEmpty { return custom }
  return titleCase(content.label)
}
